utilities/bulk-string-replace.py

```diff
@@ -0,0 +1,179 @@
+    moving many prefixes from a node to another. Instead of manually update
+    prefix nodes and descriptions, this tool replaces all instances of a string
+    with another.
+    
+    Currently the following attributes are checked for strings to replace:
+    * node
+    * description
+
+    Usage: bulk-string-replace.py pattern replacement
+
+    When the script is executed, it begins by fetching all prefixes from NIPAP
+    which matches the pattern. It then iterates over them and prints out a
+    "diff", showing what lines matches and what they would be changed to. Above
+    each entry a number is displayed which identifies the entry.
+
+    Then the user is asked to select what changes to perform. The selection is
+    done by entering the numbers identifying the entries to change. Three types
+    of input is supported:
+    * The string 'all' - perform all listed replacements
+    * A comma separated list of numbers (eg. 5,7,12,37).
+      Replacement will be performed on the listed entries.
+    * A comma separated list of numbers prefixed with an exclamation mark (!, eg
+      !5,7,12,37). Replacement will be performed on all entries EXCEPT the
+      listed entries.
+"""
+
+import ConfigParser
+import os
+import time
+import re
+import sys
+
+from pynipap import Prefix, Pool, VRF
+import pynipap
+
+BATCH_SIZE = 100
+
+def replace(pattern, replacement):
+
+    # can we use colors?
+    use_colors = sys.stdout.isatty()
+
+    # Fetch prefixes matching the string to replace
+    print "Fetching prefixes from NIPAP... ",
+    sys.stdout.flush()
+    n = 1
+    prefix_list = []
+    t0 = time.time()
+    query = {
+        'operator': 'or',
+        'val1': {
+            'operator': 'regex_match',
+            'val1': 'description',
+            'val2': pattern
+        },
+        'val2': {
+            'operator': 'regex_match',
+            'val1': 'node',
+            'val2': pattern
+        }
+    }
+    full_result = Prefix.search(query, { 'parents_depth': -1, 'max_result': BATCH_SIZE })
+    prefix_result = full_result['result']
+    prefix_list += prefix_result
+    print len(prefix_list), 
+    sys.stdout.flush()
+    while len(prefix_result) == 100:
+        full_result = Prefix.smart_search(pattern, { 'parents_depth': -1, 'max_result': BATCH_SIZE, 'offset': n * BATCH_SIZE })
+        prefix_result = full_result['result']
+        prefix_list += prefix_result
+        print len(prefix_list), 
+        sys.stdout.flush()
+        n += 1
+
+    t1 = time.time()
+    print " done in %.1f seconds" % (t1 - t0)
+
+    # Display list
+    print_pattern = "%-2s%-14s%-2s%-30s%-20s%s"
+    print "\n\nPrefixes to change:"
+    print print_pattern % ("", "VRF", "", "Prefix", "Node", "Description")
+    i_match = 0
+    for i, prefix in enumerate(prefix_list):
+        if prefix.match:
+            if use_colors:
+                print "\x1b[0m",
+            print " -- %d --" % i
+            color = "\x1b[91m"
```

Comment:
Does this really honour use_colors? You use it just above but here we set the color variable which is used on line 96 to print stuff. I don't know color codes by heart. Aren't there names for these in some lib or can we define them somewhere instead?